### PR TITLE
Plugins support work

### PR DESCRIPTION
**Pull request by est77**
- Added a plugin cache, to avoid loading the same plugin multiple times.
- Plugins can now have optional initialization / uninitialization functions.
- Modified the Display class to use the new plugin cache to load plugins.
- Improvements to Display class error messages and other small fixes.

### CODE REVIEW

File: src/appleseed/renderer/utility/plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+//
+
+// Interface header.
+#include "plugin.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+
+// appleseed.foundation headers.
+#include <foundation/platform/sharedlibrary.h>
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.cpp]: `#include "foundation/platform/sharedlibrary.h"`

File: src/appleseed/renderer/utility/plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+namespace renderer
+{
+
+//
+// ExceptionPluginInitializationFailed class implementation.
+//
+
+ExceptionPluginInitializationFailed::ExceptionPluginInitializationFailed()
+  : Exception()
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.cpp]: It's not necessary to explicitely call the base constructor here.

File: src/appleseed/renderer/utility/plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    explicit Impl(shared_ptr<SharedLibrary> p)
+      : m_library(p)
+    {
+        assert(m_library);
+    }
+
+    shared_ptr<SharedLibrary> m_library;
+};
+
+Plugin::Plugin(Impl *i)
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.cpp]: `Impl* impl`

File: src/appleseed/renderer/utility/plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        if (shared_ptr<SharedLibrary> lib = it->second.lock())
+        {
+            Plugin::Impl *i = new Plugin::Impl(lib);
+            return auto_release_ptr<Plugin>(new Plugin(i));
+        }
+    }
+
+    // If this plugin is not in the cache, load the shared lib.
+    shared_ptr<SharedLibrary> lib(
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.cpp]: Maybe on one line?

File: src/appleseed/renderer/utility/plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        reinterpret_cast<Plugin::InitPluginFnType>(
+            lib->get_symbol("initialize_plugin"));
+
+    if (init_fn)
+    {
+        if (!init_fn())
+            throw ExceptionPluginInitializationFailed();
+    }
+
+    Plugin::Impl *i = new Plugin::Impl(lib);
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.cpp]: `Plugin::Impl* impl`. Reserve `i` for iterators.

File: src/appleseed/renderer/utility/plugin.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+#define APPLESEED_RENDERER_UTILITY_PLUGIN_H
+
+// appleseed.renderer headers.
+
+// appleseed.foundation headers.
+#include "foundation/core/concepts/noncopyable.h"
+#include "foundation/core/exceptions/exception.h"
+#include "foundation/utility/autoreleaseptr.h"
+
+// Standard headers.
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: Remove.

File: src/appleseed/renderer/utility/plugin.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    // Constructor.
+    ExceptionPluginInitializationFailed();
+};
+
+
+class Plugin
+  : public foundation::NonCopyable
+{
+  public:
+
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: Remove blank line.

File: src/appleseed/renderer/utility/plugin.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_RENDERER_UTILITY_PLUGIN_H
+#define APPLESEED_RENDERER_UTILITY_PLUGIN_H
+
+// appleseed.renderer headers.
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: Remove.

File: src/appleseed/renderer/utility/plugin.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    void release();
+
+    // Get a symbol from the plugin.
+    void* get_symbol(const char* name, const bool no_throw = true) const;
+
+  private:
+    friend class PluginCache;
+
+    struct Impl;
+    Impl *impl;
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: `Impl* impl`

File: src/appleseed/renderer/utility/plugin.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    // Constructor.
+    explicit Plugin(Impl *i);
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: `Impl* impl`

File: src/appleseed/renderer/utility/plugin.h [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    // Destructor.
+    ~Plugin();
+};
+
+
+class PluginCache
+{
+  public:
+    static foundation::auto_release_ptr<Plugin> load(const char *path);
```
Comment (dictoon) [inline src/appleseed/renderer/utility/plugin.h]: Position of `*`.

File: src/appleseed/renderer/modeling/display/display.cpp [hunk 12]
```diff
@@ -118,7 +119,7 @@ void Display::open(const Project& project)
     }
     catch (const ExceptionDictionaryItemNotFound&)
     {
-        RENDERER_LOG_FATAL("%s", "cannot open display: bad parameters.");
+        RENDERER_LOG_FATAL("%s", "cannot open display: missing plugin_name parameter.");
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/display/display.cpp]: `RENDERER_LOG_FATAL()` will terminate the application. It should almost never be called, except in command line tools themselves (not in library code).

Change `open()` to return some kind of success/failure result (even a simple bool) and handle errors in the caller.

**Comment by dictoon** [comment]
Pretty cool!

**Comment by Narann** [comment]
Yeah, it sounds very good! :D

**Pull request merged**